libvmaf/src/feature/funque_adm.c

```diff
@@ -0,0 +1,319 @@
+
+    // Mirror last `reflect` columns
+    for (j = 0; j != reflect; j++)
+      dest[i * out_width + out_width - reflect + j] = dest[i * out_width + out_width - reflect - 2 - j];
+  }
+
+  // Mirror first `reflect` and last `reflect` rows
+  for (i = 0; i != reflect; i++)
+  {
+    memcpy(&dest[(reflect - 1) * out_width - i * out_width], &dest[reflect * out_width + (i + 1) * out_width], sizeof(funque_dtype) * out_width);
+    memcpy(&dest[(out_height - reflect) * out_width + i * out_width], &dest[(out_height - reflect - 1) * out_width - (i + 1) * out_width], sizeof(funque_dtype) * out_width);
+  }
+}
+
+void integral_image_adm(const funque_dtype *src, size_t width, size_t height, funque_dtype *sum)
+{
+  int i, j;
+  for (i = 0; i < (height + 1); ++i)
+  {
+    for (j = 0; j < (width + 1); ++j)
+    {
+      if (i == 0 || j == 0)
+        continue;
+
+      funque_dtype val = src[(i - 1) * width + (j - 1)];
+
+      if (i >= 1)
+      {
+        val += sum[(i - 1) * (width + 1) + j];
+        if (j >= 1)
+        {
+          val += sum[i * (width + 1) + j - 1] - sum[(i - 1) * (width + 1) + j - 1];
+        }
+      }
+      else
+      {
+        if (j >= 1)
+        {
+          val += sum[i * width + j - 1];
+        }
+      }
+      sum[i * (width + 1) + j] = val;
+    }
+  }
+}
+
+void integral_image_adm_sums(funque_dtype *x, int k, int stride, funque_dtype *mx, int width, int height)
+{
+  funque_dtype *x_pad, *int_x;
+  int i, j;
+
+  int x_reflect = (int)((k - stride) / 2);
+
+  x_pad = (funque_dtype *)malloc(sizeof(funque_dtype) * (width + (2 * x_reflect)) * (height + (2 * x_reflect)));
+  
+  reflect_pad_adm(x, width, height, x_reflect, x_pad);
+  
+  size_t r_width = width + (2 * x_reflect);
+  size_t r_height = height + (2 * x_reflect);
+
+  int_x = (funque_dtype *)calloc((r_width + 1) * (r_height + 1), sizeof(funque_dtype));
+
+  integral_image_adm(x_pad, r_width, r_height, int_x);
+
+  for (i = 0; i < height; i++)
+  {
+    for (j = 0; j < width; j++)
+    {
+      mx[i * width + j] = (int_x[i * (width + 3) + j] - int_x[i * (width+3) + j + k] - int_x[(i + k) * (width+3) + j] + int_x[(i + k) *(width+3) + j + k]);
+    }
+  }
+  free(x_pad);
+  free(int_x);
+}
+
+void dlm_decouple(dwt2buffers ref, dwt2buffers dist, dwt2buffers dlm_rest, dwt2buffers dlm_add)
+{
+  funque_dtype eps = 1e-30;
+  size_t width = ref.width;
+  size_t height = ref.height;
+  int i, j, k, index;
+
+  funque_dtype *psi_ref = (funque_dtype *)calloc(width * height, sizeof(funque_dtype));
+  funque_dtype *psi_dist = (funque_dtype *)calloc(width * height, sizeof(funque_dtype));
+  funque_dtype *psi_diff = (funque_dtype *)calloc(width * height, sizeof(funque_dtype));
+  funque_dtype *var_k;
+  funque_dtype val;
+  funque_dtype tmp_val;
+
+  for (i = 0; i < height; i++)
+  {
+    for (j = 0; j < width; j++)
+    {
+      index = i * width + j;
+#if FUNQUE_DOUBLE_DTYPE
```

Comment:
As we discussed during the meeting, we should replace the use of `atan` with the cosine-based method. Please refer to https://github.com/Netflix/vmaf/blob/master/libvmaf/src/feature/adm_tools.c#L161.

Reply:
based on the recent test with cosine-based method we have observed maximum difference of 0.007 in corner case, we will mark this issue as closed and can be revisited later on need basis.